### PR TITLE
Create "Highlight mentions" addon

### LINKED ISSUE #6185: Highlight mentions
### Why this would be helpful

This would make it easier to see mentions of your username.

### How this addon works

This addon would highlight mentions of your username in comments and forum posts.
A comment with two replies, one of them is a reply to me, my username is highlighted in orange

GitHub has this feature built in.

### Possible alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mybearworld**
Resolves #6185

### Changes

Add "Highlight mentions" addon.

### Reason for changes

This would make it easier to see mentions of your username.

### Tests

Tested on Edge.

### CODE REVIEW

File: addons/mentions/addon.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      "id": "direct",
+      "type": "boolean",
+      "default": false
+    }
+  ],
+  "credits": [
+    {
+      "name": "mybearworld",
+      "link": "https://scratch.mit.edu/users/mybearworld"
+    }
```
Comment (DNin01) [inline addons/mentions/addon.json:30]:
```suggestion
  "customCssVariables": [
    {
      "name": "button-text",
      "value": {
        "type": "textColor",
        "black": "#000",
        "white": "#fff",
        "source": {
          "type": "settingValue",
          "settingId": "color"
        }
      }
    }
  ],
  "userscripts": [
    {
      "matches": ["projects", "studios", "profiles", "forums"],
      "url": "userscript.js"
    }
  ],
  "userstyles": [
    {
      "matches": ["projects", "studios", "profiles", "forums"],
      "url": "userstyle.css"
    }
  ],
  "settings": [
    {
      "name": "Highlight color",
      "id": "color",
      "type": "color",
      "default": "#855cd6"
    },
    {
      "name": "Highlight direct links",
      "id": "direct",
      "type": "boolean",
      "default": false
    }
  ],
  "credits": [
    {
      "name": "mybearworld",
      "link": "https://scratch.mit.edu/users/mybearworld"
    },
    {
      "name": "DNin01",
      "link": "https://github.com/DNin01"
    }
```

File: addons/mentions/userstyle.css [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.sa-mention {
+  background-color: var(--darkWww-button, #4d97ff);
+  color: var(--darkWww-button-text, white) !important;
```
Comment (DNin01) [inline addons/mentions/userstyle.css:3]:
```suggestion
  background-color: var(--mentions-color);
  color: var(--mentions-button-text) !important;
```

**Review by DNin01 — COMMENTED** [review]
I think it would be nice to add the ability to change the color and update the default color to the purple that Scratch will have switched to by the time we release this addon, I'm assuming. I've proposed these changes in the code review comments below.

(Edits from maintainers are disabled so I need your permission to make edits. You might have disabled this because it said that it would give us access to secrets. What it's saying here is that if maintainers are modifying workflows in your fork, they could reveal the values of secrets and gain access to other branches on the fork. If you trust the maintainers, it's safe to allow edits and "access to secrets", if you want to, just so you know. Regardless of whether you allow edits, we can still propose suggestions that you can choose to apply.)

File: addons/mentions/addon.json (hunk 1 above)
Comment (mybearworld) [reply addons/mentions/addon.json:30]:
```suggestion
"customCssVariables": [
  {
    "name": "button-text",
    "value": {
      "type": "textColor",
      "black": "#000",
      "white": "#fff",
      "source": {
        "type": "settingValue",
        "settingId": "color"
      }
    }
  }
],
"userscripts": [
  {
    "matches": ["projects", "studios", "profiles", "forums"],
    "url": "userscript.js"
  }
],
"userstyles": [
  {
    "matches": ["projects", "studios", "profiles", "forums"],
    "url": "userstyle.css"
  }
],
"settings": [
  {
    "name": "Highlight color",
    "id": "color",
    "type": "color",
    "default": "#855cd6"
  },
  {
    "name": "Highlight direct links",
    "id": "direct",
    "type": "boolean",
    "default": false
  }
],
"credits": [
  {
    "name": "mybearworld",
    "link": "https://scratch.mit.edu/users/mybearworld"
  },
  {
    "name": "DNin01",
    "link": "https://github.com/DNin01"
  }
]
```

File: addons/mentions/userstyle.css (hunk 2 above)
Comment (mybearworld) [reply addons/mentions/userstyle.css]: Thank you for letting me know about the issue with disallowing edits, I've allowed them now - I just didn't want to risk anything.

File: addons/mentions/addon.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  ],
+  "settings": [
+    {
+      "name": "Highlight color",
+      "id": "color",
+      "type": "color",
+      "default": "#855cd6"
+    },
+    {
+      "name": "Highlight direct links",
```
Comment (DNin01) [inline addons/mentions/addon.json:40]: Might want to attempt to clarify this setting. I didn't know what it did until I enabled it on a forum post, and I'm still not 100% sure what it does.

Making a concise name for this setting is hard. #5320 could help here.

**Review by DNin01 — APPROVED** [review]
Tested on Edge 114.

File: addons/mentions/addon.json (hunk 3 above)
Comment (mybearworld) [reply addons/mentions/addon.json:40]: It matches `https://scratch.mit.edu/users/mybearworld` links as opposed to just `/users/mybearworld` links. This should probably just be a "highlight in forums" option.

**Comment by mxmou** [comment]
Non-theme addons generally shouldn't have color settings. Most people won't change the color and it makes it impossible to change it automatically based on dark mode settings. See #1262

**Comment by DNin01** [comment]
> Non-theme addons generally shouldn't have color settings. Most people won't change the color and it makes it impossible to change it automatically based on dark mode settings. See #1262

Non-theme addons such as `editor-stepping` and `paint-snap` have customizable colors.

I was thinking of having this option mainly to make it possible to choose between Scratch's default blue/purple color or a shade of yellow/orange, like the one used for mentions on GitHub. But if the goal is not to have too many customizations, sure, I guess we don't need it, but we should choose between a shade of yellow/orange or Scratch's default colors.

**Comment by Secret-chest** [comment]
I want to control my colours.

**Comment by mxmou** [comment]
> I was thinking of having this option in `mentions` mainly to make it possible to choose between Scratch's default **blue/purple** color or a shade of **yellow/orange**, like the one used for mentions on GitHub. But if the goal is not to have too many customizations, sure, I guess we don't need it, but we should choose between a shade of yellow/orange or Scratch's default colors.

If that was the intention, could an orange color from Scratch's color palette (maybe the Variables color - `#FF8C1A`) be the default?

**Comment by mybearworld** [comment]
> > I was thinking of having this option in `mentions` mainly to make it possible to choose between Scratch's default **blue/purple** color or a shade of **yellow/orange**, like the one used for mentions on GitHub. But if the goal is not to have too many customizations, sure, I guess we don't need it, but we should choose between a shade of yellow/orange or Scratch's default colors.
> 
> If that was the intention, could an orange color from Scratch's color palette (maybe the Variables color - `#FF8C1A`) be the default?

The colors will probably have changed by the time of the release of the next version, so the color would be `#ffa54c`.

**Comment by mxmou** [comment]
> > > I was thinking of having this option in `mentions` mainly to make it possible to choose between Scratch's default **blue/purple** color or a shade of **yellow/orange**, like the one used for mentions on GitHub. But if the goal is not to have too many customizations, sure, I guess we don't need it, but we should choose between a shade of yellow/orange or Scratch's default colors.
> > 
> > 
> > If that was the intention, could an orange color from Scratch's color palette (maybe the Variables color - `#FF8C1A`) be the default?
> 
> The colors will probably have changed by the time of the release of the next version, so the color would be `#ffa54c`.

The original colors will still be the default.

**Comment by mybearworld** [comment]
> > > > I was thinking of having this option in `mentions` mainly to make it possible to choose between Scratch's default **blue/purple** color or a shade of **yellow/orange**, like the one used for mentions on GitHub. But if the goal is not to have too many customizations, sure, I guess we don't need it, but we should choose between a shade of yellow/orange or Scratch's default colors.
> > > 
> > > 
> > > If that was the intention, could an orange color from Scratch's color palette (maybe the Variables color - `#FF8C1A`) be the default?
> > 
> > 
> > The colors will probably have changed by the time of the release of the next version, so the color would be `#ffa54c`.
> 
> The original colors will still be the default.

Oh, yeah, I completely forgot about that

File: addons/mentions/addon.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      "matches": ["projects", "studios", "profiles", "forums"],
+      "url": "userstyle.css"
+    }
+  ],
+  "settings": [
+    {
+      "name": "Highlight color",
+      "id": "color",
+      "type": "color",
+      "default": "#855cd6"
```
Comment (DNin01) [inline addons/mentions/addon.json:37]: Should we change this color to a shade of yellow/orange by default? Should we include two presets so the user can quickly switch it between yellow/orange and purple? Should we remove the setting entirely?

Comment (mybearworld) [reply addons/mentions/addon.json:37]: It might make sense to just have it be a native Scratch color for more easy compatibility with dark-www

Comment (Samq64) [reply addons/mentions/addon.json:37]: The orange used for the message count is customizable with `dark-www` now.

File: addons/mentions/userscript.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+export default async function ({ addon }) {
+  if (!(await addon.auth.fetchIsLoggedIn())) {
```
Comment (Samq64) [inline addons/mentions/userscript.js:2]: Note that `scratch-www` allows dynamic logins.

File: addons/mentions/userscript.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    return;
+  }
+  const username = await addon.auth.fetchUsername();
+  setInterval(() => {
```
Comment (Samq64) [inline addons/mentions/userscript.js:6]: This should probably use `waitForElement` instead.

File: addons/mentions/userstyle.css [hunk 7]
```diff
@@ -0,0 +1,12 @@
+.sa-mention {
+  background-color: var(--mentions-color);
```
Comment (Samq64) [inline addons/mentions/userstyle.css:2]: As mentioned above the setting can be removed entirely now that dark-www allows customizing the orange badge colour.
```suggestion
  background-color: var(--darkWww-messageIndicatorColor, #ffab1a);
```

File: addons/mentions/userstyle.css [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  color: var(--mentions-button-text) !important;
```
Comment (Samq64) [inline addons/mentions/userstyle.css:3]:
```suggestion
  color: var(--darkWww-messageIndicatorColorText, #fff) !important;
```

File: addons/mentions/userscript.js (hunk 5 above)
Comment (DNin01) [reply addons/mentions/userscript.js:2]: (Logging in without having to reload the page)

File: addons/mentions/userstyle.css (hunk 7 above)
Comment (mybearworld) [reply addons/mentions/userstyle.css:2]: This looks pretty bad by default...
image

Comment (mybearworld) [reply addons/mentions/userstyle.css:2]: The button color probably works still
image

File: addons/mentions/addon.json (hunk 3 above)
Comment (Samq64) [reply addons/mentions/addon.json:24]: Maybe
```suggestion
      "name": "Highlight full links",
```

File: addons/addons.json [hunk 9]
```diff
@@ -157,6 +157,7 @@
   "asset-conflict-dialog",
   "remaining-replies",
   "forum-user-agent",
+  "mentions",
```
Comment (Samq64) [inline addons/addons.json:160]: Maybe this could be something like `highlight-mentions` since I have a mentions addon PR open as well?

**Review by Samq64 — COMMENTED** [review]
> This should probably use `waitForElement` instead.

~~I see now what you did it and there was nothing wrong with it. `waitForElement` breaks re-enabling the setting without a page reload and it's easiest to just revert that commit to fix it.~~

Why not just use CSS directly?

**Comment by Samq64** [comment]
@mybearworld deleting your fork caused all your PRs to be closed. Was that intentional?

**Comment by mybearworld** [comment]
> @mybearworld deleting your fork caused all your PRs to be closed. Was that intentional?

Oh! Thank you for letting me know! I meant to delete some other forks, not this one.

**Comment by mybearworld** [comment]
I accidentally deleted my fork. New PR is at #7971

**Pull request closed without merging**